File: src/main/java/org/hotamachisubaru/miniutility/Listener/NameColor.java
package org.hotamachisubaru.miniutility.Listener;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;

import java.util.HashMap;
import java.util.Map;

public class NameColor implements Listener {
    private static final Map<Player, Boolean> waitingForColorInput = new HashMap<>();

    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (event.getClickedInventory() == null || event.getCurrentItem() == null) return;

        Player player = (Player) event.getWhoClicked();

        // 名前の色変更アイテムをクリックしたか判定
        if (event.getCurrentItem().getType() == Material.GREEN_DYE) {
            event.setCancelled(true);
            promptForColorInput(player);
        }
    }

    private void promptForColorInput(Player player) {
        waitingForColorInput.put(player, true);
        player.sendMessage("チャットにカラーコードを入力してください（例：&6）。");
        player.closeInventory();
    }

    public static boolean isWaitingForColorInput(Player player) {
        return waitingForColorInput.getOrDefault(player, false);
    }

    public static void setWaitingForColorInput(Player player, boolean waiting) {
        waitingForColorInput.put(player, waiting);
    }
}

File: src/main/java/org/hotamachisubaru/miniutility/Listener/NameColor.kt
package org.hotamachisubaru.miniutility.Listener

import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent

class NameColor : Listener {
    @EventHandler
    fun onPlayerChat(event: AsyncPlayerChatEvent) {
        val player = event.player

        // プレイヤーがカラーコードの入力待ち状態であるか確認
        if (waitingForColorInput.getOrDefault(player, false)) {
            event.isCancelled = true // カラーコードの入力を通常のチャットメッセージとしては処理しない

            val message = event.message

            // カラーコードの形式が正しいかを確認（例：&6など）
            if (message.matches("^&[0-9a-fA-F]$".toRegex())) {
                // プレイヤー名を指定されたカラーコードで設定（ニックネームが既に存在する場合はそれに色をつける）
                val currentNickname = player.displayName
                val coloredName = ChatColor.translateAlternateColorCodes('&', message) + ChatColor.stripColor(currentNickname)
                player.setDisplayName(coloredName)
                player.setPlayerListName(coloredName)
                player.sendMessage(ChatColor.GREEN.toString() + "プレイヤー名が " + message + " の色に変更されました！")
            } else {
                player.sendMessage(ChatColor.RED.toString() + "無効なカラーコードです。正しい形式で入力してください（例：&6）。")
            }

            // 入力待ち状態をリセット
            waitingForColorInput[player] = false
        }
    }

    companion object {
        // プレイヤーがカラーコードの入力を待っているかどうかを記録するマップ
        var waitingForColorInput: MutableMap<Player, Boolean> = HashMap()
    }
}
